extended/src/main/java/io/kubernetes/client/extended/kubectl/util/deployment/DeploymentHelper.java

```diff
@@ -0,0 +1,134 @@
+    List<V1ReplicaSet> rsList = listReplicaSets(deployment, api);
+    V1ReplicaSet newRs = findNewReplicaSet(deployment, rsList);
+    findOldReplicaSets(rsList, newRs, oldRSes, allOldRSes);
+    return newRs;
+  }
+
+  public static Long Revision(V1ObjectMeta meta) {
+    String v = meta.getAnnotations().get(REVISION_ANNOTATION);
+    if (v == null) return 0L;
+    return Long.valueOf(v);
+  }
+
+  private static List<V1ReplicaSet> listReplicaSets(V1Deployment deployment, AppsV1Api api)
+      throws ApiException {
+    String namespace = deployment.getMetadata().getNamespace();
+    LabelSelector selector = LabelSelector.parse(deployment.getSpec().getSelector());
+    List<V1ReplicaSet> all = rsListFromClient(namespace, selector.toString(), api);
+    List<V1ReplicaSet> owned = new ArrayList<>(all.size());
+    for (V1ReplicaSet rs : all) {
+      List<V1OwnerReference> refs = rs.getMetadata().getOwnerReferences();
+      Optional<V1OwnerReference> ref =
+          refs.stream().filter(o -> o.getController() != null && o.getController()).findAny();
+      if (ref.isPresent() && ref.get().getUid().equals(deployment.getMetadata().getUid())) {
+        owned.add(rs);
+      }
+    }
+    return owned;
+  }
+
+  private static List<V1ReplicaSet> rsListFromClient(
+      String namespace, String selector, AppsV1Api api) throws ApiException {
+    V1ReplicaSetList rsList =
+        api.listNamespacedReplicaSet(
+            namespace, null, null, null, null, selector, null, null, null, null, null);
+    return rsList.getItems();
+  }
+
+  private static V1ReplicaSet findNewReplicaSet(
+      V1Deployment deployment, List<V1ReplicaSet> rsList) {
+    rsList.sort(
+        (o1, o2) -> {
+          if (o1.getMetadata()
+              .getCreationTimestamp()
+              .equals(o2.getMetadata().getCreationTimestamp())) {
+            return o1.getMetadata().getName().compareTo(o2.getMetadata().getName());
+          }
+          if (o1.getMetadata()
+              .getCreationTimestamp()
+              .isBefore(o2.getMetadata().getCreationTimestamp())) {
+            return -1;
+          }
+          return 1;
+        });
+    for (V1ReplicaSet rs : rsList) {
+      if (equalIgnoreHash(rs.getSpec().getTemplate(), deployment.getSpec().getTemplate())) {
+        return rs;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * EqualIgnoreHash returns true if two given podTemplateSpec are equal, ignoring the diff in value
+   * of Labels[pod-template-hash] We ignore pod-template-hash because: 1. The hash result would be
+   * different upon podTemplateSpec API changes (e.g. the addition of a new field will cause the
+   * hash code to change) 2. The deployment template won't have hash labels
+   */
+  private static boolean equalIgnoreHash(V1PodTemplateSpec template1, V1PodTemplateSpec template2) {
+    // todo : use deep copy
+    String hash1 = template1.getMetadata().getLabels().get(DEFAULT_DEPLOYMENT_UNIQUE_LABEL_KEY);
```

Comment:
I would refactor this a little:

```
if (!Objects.equals(template1.getSpec(), template2.getSpec())) {
    return false;
}
return equalIgnoreHash(template1.getMetadata(), template2.getMetadata());
```

That way you only do the munging if it's really truly necessary, rather than all the time.

Reply:
Good suggestion, already refactored.